yasa/heart.py

```diff
@@ -0,0 +1,177 @@
+
+__all__ = ['hrv_stage']
+
+
+def hrv_stage(data, sf, *, hypno=None, include=(2, 3, 4), threshold="2min", equal_length=False,
+              rr_limit=(400, 2000), verbose=False):
+    """Calculate heart rate and heart rate variability (HRV) features from an ECG.
+
+    By default, the cardiac features are calculated for each period of N2, N3 or REM sleep that
+    are longer than 2 minutes.
+
+    .. versionadded:: 0.6.2
+
+    Parameters
+    ----------
+    data : :py:class:`numpy.ndarray`
+        Single-channel ECG data. Must be a 1D NumPy array.
+    sf : float
+        The sampling frequency of the data.
+    hypno : array_like
+        Sleep stage (hypnogram). The heart rate calculation will be applied for each sleep stage
+        defined in ``include`` (default = N2, N3 and REM sleep separately).
+
+        The hypnogram must have the same number of samples as ``data``.
+        To upsample your hypnogram, please refer to
+        :py:func:`yasa.hypno_upsample_to_data`.
+
+        .. note::
+            The default hypnogram format in YASA is a 1D integer
+            vector where:
+
+            - -2 = Unscored
+            - -1 = Artefact / Movement
+            - 0 = Wake
+            - 1 = N1 sleep
+            - 2 = N2 sleep
+            - 3 = N3 sleep
+            - 4 = REM sleep
+    include : tuple, list or int
+        Values in ``hypno`` that will be included in the mask. The default is
+        (2, 3, 4), meaning that the detection is applied on N2, N3 and REM
+        sleep separately.
+    threshold : str
+        Only periods of a given stage that exceed the duration defined in ``threshold`` will be
+        kept in subsequent analysis. The default is 2 minutes ('2min'). Other possible values
+        include: '5min', '15min', '30sec', '1hour', etc. To disable thresholding, use '0min'.
+    equal_length : bool
+        If True, the periods will all have the exact duration defined in ``threshold``.
+        That is, periods that are longer than the duration threshold will be divided into
+        sub-periods of exactly the length of threshold.
+    rr_limit : tuple
+        Lower and upper limit for the RR interval. Default is 400 to 2000 ms, corresponding to a
+        heart rate of 30 to 150 bpm. RR intervals outside this range will be set to NaN and
+        filled with linear interpolation. Use ``rr_limit=(0, np.inf)`` to disable RR correction.
+    verbose : bool or str
+        Verbose level. Default (False) will only print warning and error
+        messages. The logging levels are 'debug', 'info', 'warning', 'error',
+        and 'critical'. For most users the choice is between 'info'
+        (or ``verbose=True``) and warning (``verbose=False``). Set this to True if you are getting
+        invalid results and want to better understand what is happening.
+
+    Returns
+    -------
+    epochs : :py:class:`pandas.DataFrame`
+        Output dataframe with stage and epoch number as index. The columns are
```

Comment:
Two thoughts:

1. Here it says the indices are `stage` and `epoch` but it returns with names `values` and `epoch`. I'd suggest either clarifying to the user knows what to expect, or just renaming it here or in the ouput.
2. I'm not sure it should be returned with indices at all... In my experience, people vary a lot in there comfort/familiarity with using pandas indices, and in some cases a "regular" dataframe is easier to digest. Users who like taking advantage of indices can easily set them themselves. I guess you are thinking it makes the connection to the `rpeaks` dictionary keys more explicit, which is definitely an upside, so I see why you would want to keep it this way. But I'll also say I found the tuples-as-keys a bit confusing for `rpeaks` as well, another option might be a nested dictionary (epochs within stages). Keeping it same or changing it are both okay with me!

Reply:
Thanks for catching 1!

As for 2 🤔 ... I do think that (multi)-index in Pandas are super powerful, and that sadly most Pandas users do not really take advantage of them. There are several functions in YASA that return a dataframe with multi-index, e.g.:
- yasa.bandpower
- detection summary aggregated by stage and channels, e.g. `sw.summary(grp_chan=True, grp_stage=True)`

Therefore, this kind of output is not completely out-of-the-blue for YASA users. It also make things easier with the `rpeaks` dictionary indexing as you noted. My first implementation was actually with a nested dict, but I think I like the tuples-as-keys version better.

Reply:
I wholeheartedly agree, multi-index is powerful and likely underused. Pandas obviously _wants_ us to use them, since otherwise your code gets filled with `.reset_index(drop=False)` everywhere lol. After submitting this review it occurred to me how YASA has multi-index elsewhere, so I figured this was indeed a conscious decision. You've got my full support to push multi-indices on people! :)